### PR TITLE
Leaderboard query

### PR DESCRIPTION
**Pull request by mrsmkl**
Adds the following
* leaderboard query
* reading validator history
* read contract addresses from registry
* reading extra logs to get the epoch rewards (outdated)
* counting attestation events

### CODE REVIEW

File: apps/ethereum_jsonrpc/lib/ethereum_jsonrpc.ex [hunk 1]
```diff
@@ -343,10 +347,10 @@ defmodule EthereumJSONRPC do
   @spec json_rpc(Transport.batch_request(), json_rpc_named_arguments) ::
           {:ok, Transport.batch_response()} | {:error, reason :: term()}
   def json_rpc(request, named_arguments) when (is_map(request) or is_list(request)) and is_list(named_arguments) do
+    # IO.inspect(request)
```
Comment (nambrot) [inline apps/ethereum_jsonrpc/lib/ethereum_jsonrpc.ex]: Remove

File: apps/ethereum_jsonrpc/lib/ethereum_jsonrpc/contract.ex [hunk 2]
```diff
@@ -52,6 +52,9 @@ defmodule EthereumJSONRPC.Contract do
       end
       |> Enum.into(%{}, &{&1.id, &1})
 
+    # IO.inspect(requests)
```
Comment (nambrot) [inline apps/ethereum_jsonrpc/lib/ethereum_jsonrpc/contract.ex]: Remove

File: apps/explorer/lib/explorer/celo/account_reader.ex [hunk 3]
```diff
@@ -9,6 +9,7 @@ defmodule Explorer.Celo.AccountReader do
 
   def account_data(%{address: account_address}) do
     data = fetch_account_data(account_address)
+    # IO.inspect(data)
```
Comment (nambrot) [inline apps/explorer/lib/explorer/celo/account_reader.ex]: Remove

File: apps/explorer/lib/explorer/celo/account_reader.ex [hunk 4]
```diff
@@ -170,12 +195,51 @@ defmodule Explorer.Celo.AccountReader do
...
+
+    methods = [
+      %{
+        contract_address: "0x000000000000000000000000000000000000ce10",
+        function_name: "getAddressForString",
+        args: [name]
+      }
+    ]
+
+    # IO.inspect(methods)
```
Comment (nambrot) [inline apps/explorer/lib/explorer/celo/account_reader.ex]: Removee

File: apps/explorer/lib/explorer/chain.ex [hunk 5]
```diff
@@ -3654,6 +3654,32 @@ defmodule Explorer.Chain do
...
 
+  def query_leaderboard do
+    result =
+      SQL.query(Repo, """
+        SELECT competitors.address, b.name, SUM(rate*value+fetched_coin_balance+celo_account.locked_gold)*multiplier AS score
+        FROM addresses, exchange_rates, competitors, claims, celo_account, celo_account as b,
+          (SELECT claims.claimed_address AS address, COALESCE(SUM(value),0) AS value
+          FROM address_current_token_balances, claims
+          WHERE address_hash=claims.claimed_address
+          AND token_contract_address_hash='\\x88f24de331525cf6cfd7455eb96a9e4d49b7f292'
```
Comment (nambrot) [inline apps/explorer/lib/explorer/chain.ex]: What are these constants?

File: apps/explorer/lib/explorer/token/balance_reader.ex [hunk 6]
```diff
@@ -42,12 +42,15 @@ defmodule Explorer.Token.BalanceReader do
...
          token_contract_address_hash: token_contract_address_hash
        }) do
-    %{
+    res = %{
       contract_address: token_contract_address_hash,
       function_name: "balanceOf",
       args: [address_hash],
-      block_number: block_number
+      block_number: block_number,
+      gasprice: "1000000000000000000"
```
Comment (nambrot) [inline apps/explorer/lib/explorer/token/balance_reader.ex]: Why gasprice here?

File: apps/explorer/lib/explorer/token/balance_reader.ex [hunk 7]
```diff
@@ -42,12 +42,15 @@ defmodule Explorer.Token.BalanceReader do
          block_number: block_number,
          token_contract_address_hash: token_contract_address_hash
        }) do
-    %{
+    res = %{
```
Comment (nambrot) [inline apps/explorer/lib/explorer/token/balance_reader.ex]: The res seems unneecessary?

File: apps/explorer/priv/repo/migrations/20191011140922_create_celo_account.exs [hunk 8]
```diff
@@ -48,10 +49,33 @@ defmodule Explorer.Repo.Migrations.CreateCeloAccount do
...
       add(:block_number, :integer, null: false)
-      add(:validator_address, :bytea, null: false)
+      add(:address, :bytea, null: false)
       timestamps(null: false, type: :utc_datetime_usec)
     end
 
-    create(index(:celo_validator_history, [:validator_address, :block_number, :index], unique: true))
+    create(index(:celo_validator_history, [:block_number, :index], unique: true))
+
+    if Mix.env() != :test do
```
Comment (nambrot) [inline apps/explorer/priv/repo/migrations/20191011140922_create_celo_account.exs]: Is this a common pattern?

File: apps/indexer/lib/indexer/transform/token_transfers.ex [hunk 9]
```diff
@@ -20,6 +20,8 @@ defmodule Indexer.Transform.TokenTransfers do
   end
 
   defp do_parse(log, %{tokens: tokens, token_transfers: token_transfers} = acc) do
+    # IO.inspect("do_parse")
```
Comment (nambrot) [inline apps/indexer/lib/indexer/transform/token_transfers.ex]: Removee

File: apps/block_scout_web/lib/block_scout_web/resolvers/address.ex [hunk 10]
```diff
@@ -12,8 +14,17 @@ defmodule BlockScoutWeb.Resolvers.Address do
 
   def get_by(_, %{hash: hash}, _) do
     case Chain.hash_to_address(hash) do
-      {:error, :not_found} -> {:error, "Address not found."}
-      {:ok, _} = result -> result
+      {:error, :not_found} ->
```
[the remark this answers is not in the document]
Comment (mrsmkl) [reply apps/block_scout_web/lib/block_scout_web/resolvers/address.ex]: No, must have been debug messages.

File: apps/explorer/priv/repo/migrations/20191011140922_create_celo_account.exs (hunk 8 above)
Comment (mrsmkl) [reply apps/explorer/priv/repo/migrations/20191011140922_create_celo_account.exs]: Let's add an issue for removing these.

File: apps/explorer/lib/explorer/chain.ex (hunk 5 above)
Comment (mrsmkl) [reply apps/explorer/lib/explorer/chain.ex]: It's the address for stable token. I guess this query will be thrown away after the stake off so it doesn't matter much that it's hard coded.

File: apps/explorer/priv/repo/migrations/20191011140922_create_celo_account.exs (hunk 8 above)
Comment (mrsmkl) [reply apps/explorer/priv/repo/migrations/20191011140922_create_celo_account.exs]: https://github.com/celo-org/celo-monorepo/issues/1846

File: apps/explorer/lib/explorer/token/balance_reader.ex (hunk 6 above)
Comment (mrsmkl) [reply apps/explorer/lib/explorer/token/balance_reader.ex]: For some reason, it complained about smaller than minimum gas price when reading from some previous blocks. Added issue here https://github.com/celo-org/celo-blockchain/issues/638

File: apps/explorer/lib/explorer/chain.ex (hunk 5 above)
Comment (nambrot) [reply apps/explorer/lib/explorer/chain.ex]: Please add a comment for what this query does, or at least a link to to documentation

File: apps/explorer/lib/explorer/chain.ex [hunk 11]
```diff
@@ -3654,6 +3654,32 @@ defmodule Explorer.Chain do
...
+        WHERE  token='\\x88f24de331525cf6cfd7455eb96a9e4d49b7f292'
+        AND claims.claimed_address = get.address
+        AND celo_account.address = addresses.hash
+        AND claims.address = competitors.address
+        AND competitors.address = b.address
+        GROUP BY competitors.address, multiplier, b.name
+        ORDER BY score
+      """)
+
+    # IO.inspect(result)
```
Comment (nambrot) [inline apps/explorer/lib/explorer/chain.ex]: Remove

File: apps/block_scout_web/lib/block_scout_web/resolvers/address.ex (hunk 10 above)
[the remark this answers is not in the document]
Comment (nambrot) [reply apps/block_scout_web/lib/block_scout_web/resolvers/address.ex]: For code reviews, having the diff littered with unnecessary changes make the review unncessarily complex and mentally strenuous. Please consider removing these before submitting a PR to review to make this easier for your reviewers

File: apps/explorer/lib/explorer/token/balance_reader.ex (hunk 6 above)
Comment (nambrot) [reply apps/explorer/lib/explorer/token/balance_reader.ex]: Please leave a comment here with a reference to the ticket

File: apps/explorer/lib/explorer/celo/account_reader.ex [hunk 12]
```diff
@@ -112,16 +114,30 @@ defmodule Explorer.Celo.AccountReader do
   end
 
   defp fetch_account_data(account_address) do
+    res =
```
Comment (nambrot) [inline apps/explorer/lib/explorer/celo/account_reader.ex]: Unnecessary res

File: apps/explorer/lib/explorer/celo/account_reader.ex [hunk 13]
```diff
@@ -131,12 +147,8 @@ defmodule Explorer.Celo.AccountReader do
...
   end
 
-  defp fetch_validators do
-    data =
-      call_methods([
-        {:validators, "currentValidators", []}
-      ])
-
+  defp fetch_validators(bn) do
+    data = call_methods([{:election, "getCurrentValidatorSigners", []}], bn)
```
Comment (nambrot) [inline apps/explorer/lib/explorer/celo/account_reader.ex]: unnecessary data

File: apps/explorer/lib/explorer/celo/account_reader.ex [hunk 14]
```diff
@@ -170,12 +194,43 @@ defmodule Explorer.Celo.AccountReader do
...
+
+    res =
+      methods
+      |> Reader.query_contracts(contract_abi)
+      |> Enum.zip(methods)
+      |> Enum.into(%{}, fn {response, %{function_name: function_name}} ->
+        {function_name, response}
+      end)
+
+    {:ok, [address]} = res["getAddressForString"]
```
Comment (nambrot) [inline apps/explorer/lib/explorer/celo/account_reader.ex]: What happens when res does not have the key?

File: apps/indexer/lib/indexer/block/fetcher.ex [hunk 15]
```diff
@@ -110,6 +111,23 @@ defmodule Indexer.Block.Fetcher do
     struct!(__MODULE__, named_arguments)
   end
 
+  #  defp append_logs(extra_logs, %{logs: logs, receipts: receipts}) do
```
Comment (nambrot) [inline apps/indexer/lib/indexer/block/fetcher.ex]: Remove

File: apps/indexer/lib/indexer/block/fetcher.ex [hunk 16]
```diff
@@ -134,15 +152,28 @@ defmodule Indexer.Block.Fetcher do
...
          {:receipts, {:ok, receipt_params}} <- {:receipts, Receipts.fetch(state, transactions_params_without_receipts)},
-         %{logs: logs, receipts: receipts} = receipt_params,
+         %{logs: tx_logs, receipts: receipts} = receipt_params,
+         logs = tx_logs ++ process_extra_logs(extra_logs),
          transactions_with_receipts = Receipts.put(transactions_params_without_receipts, receipts),
-         %{token_transfers: token_transfers, tokens: tokens} = TokenTransfers.parse(logs),
-         %{accounts: celo_accounts, validators: celo_validators, validator_groups: celo_validator_groups} =
-           CeloAccounts.parse(logs),
+         %{token_transfers: normal_token_transfers, tokens: normal_tokens} = TokenTransfers.parse(logs),
+         # %{token_transfers: fee_token_transfers, tokens: fee_tokens} =
```
Comment (nambrot) [inline apps/indexer/lib/indexer/block/fetcher.ex]: Remove

File: apps/indexer/lib/indexer/block/fetcher.ex [hunk 17]
```diff
@@ -152,6 +183,7 @@ defmodule Indexer.Block.Fetcher do
              token_transfers: token_transfers,
              transactions: transactions_with_receipts
            }),
+         #         addresses = Enum.filter(addresses_raw, fn a -> a.hash != "0x0000000000000000000000000000000000000000" end),
```
Comment (nambrot) [inline apps/indexer/lib/indexer/block/fetcher.ex]: Remove

File: apps/indexer/test/indexer/fetcher/celo_account_test.exs [hunk 18]
```diff
@@ -4,7 +4,7 @@ defmodule Indexer.Fetcher.CeloAccountsTest do
 
   import Mox
 
-  alias Explorer.Chain.{Address, Hash, CeloAccount}
+  # alias Explorer.Chain.{Address, Hash, CeloAccount}
```
Comment (nambrot) [inline apps/indexer/test/indexer/fetcher/celo_account_test.exs]: Remove

**Pull request merged**